File: aichallenge/workspace/src/aichallenge_submit/dwa/simple_dwa/src/dwa.cpp
#include "dwa.hpp"
#include "utils.hpp"
#include <iostream>
#include <cmath>
#include <limits>
#include <rclcpp/rclcpp.hpp>

DWA::DWA(const Parameters &params)
    : params_(params),
      simu_robot_(params),
      course_(params) {}

std::pair<std::vector<Path>, Path> DWA::calcInput(Robot &robot, const std::vector<Obstacle> &obstacles)
{
  // Path作成
  std::vector<Path> paths = makePath(robot);
  std::cout << "Number of paths generated: " << paths.size() << std::endl;

  // Path評価
  auto target_point = course_.getNextTargetPoint(robot.getX(), robot.getY(), robot.getTh(), params_.LOOKAHEAD_DISTANCE);
  double g_x = target_point.first;
  double g_y = target_point.second;

  auto eval_result = evalPath(paths, g_x, g_y, robot, obstacles);
  Path opt_path = eval_result.first;
  std::vector<Path> valid_paths = eval_result.second;

  // サイズ管理関数を呼び出す
  addTrajectory(opt_path, g_x, g_y, valid_paths);
  return {valid_paths, opt_path};
}

std::vector<Path> DWA::makePath(Robot &robot)
{
  // 角度と速度の範囲算出
  double min_ang_velo, max_ang_velo, min_velo, max_velo;
  // 角速度の範囲
  double range_ang_velo = params_.DT * params_.MAX_DYAWRATE;
  min_ang_velo = robot.getUTh() - range_ang_velo;
  max_ang_velo = robot.getUTh() + range_ang_velo;

  // 範囲制限
  min_ang_velo = std::max(min_ang_velo, params_.MIN_YAWRATE);
  max_ang_velo = std::min(max_ang_velo, params_.MAX_YAWRATE);
  // 速度の範囲
  double range_velo = params_.DT * params_.MAX_ACCEL;
  min_velo = robot.getUV() - range_velo;
  max_velo = robot.getUV() + range_velo;
  min_velo = std::max(min_velo, params_.MIN_SPEED);
  max_velo = std::min(max_velo, params_.MAX_SPEED);
  // 直線部分のために速度範囲を広げる
  if (robot.getUV() > params_.MIN_SPEED) {
      min_velo = std::max(min_velo, params_.MIN_SPEED);
      max_velo = std::min(max_velo, params_.MAX_SPEED);
  }

  if (min_ang_velo > max_ang_velo)
  {
    min_ang_velo = max_ang_velo - params_.YAWRATE_RESOLUTION;
  }
  if (min_velo > max_velo)
  {
    min_velo = max_velo - params_.V_RESOLUTION;
  }

  // 全てのpathのリスト
  std::vector<Path> paths;
  int count = 0;
  for (double ang_velo = min_ang_velo; ang_velo < max_ang_velo; ang_velo += params_.YAWRATE_RESOLUTION)
  {
    for (double velo = min_velo; velo < max_velo; velo += params_.V_RESOLUTION)
    {
      Path path(ang_velo, velo);
      pre_step_ = static_cast<int>(params_.PREDICT_TIME / params_.DT);
      std::vector<double> predicted = simu_robot_.predictState(ang_velo, velo, robot.getX(), robot.getY(), robot.getTh(), params_.DT, pre_step_);
      std::vector<double> next_x(predicted.begin(), predicted.begin() + pre_step_);
      std::vector<double> next_y(predicted.begin() + pre_step_, predicted.begin() + 2 * pre_step_);
      std::vector<double> next_th(predicted.begin() + 2 * pre_step_, predicted.end());

      path.setX(next_x);
      path.setY(next_y);
      path.setTh(next_th);

      paths.push_back(path);
    }
    count++;
  }
  return paths;
}

std::pair<Path, std::vector<Path>> DWA::evalPath(const std::vector<Path> &paths, double g_x, double g_y, Robot &robot, const std::vector<Obstacle> &obstacles)
{
  // 一番近い障害物判定
  std::vector<Obstacle> nearest_obs = calcNearestObs(robot, obstacles);
  std::vector<Path> valid_paths;
  std::vector<double> score_heading_angles;
  std::vector<double> score_heading_velos;
  std::vector<double> score_obstacles;

  // 全てのpathで評価を検索
  for (const auto &path : paths)
  {
    double angle_score = headingAngle(path, g_x, g_y);
    double velo_score = headingVelo(path);
    double obs_score = obstacleScore(path, nearest_obs);

    if (std::isfinite(angle_score) && std::isfinite(velo_score) && std::isfinite(obs_score))
    {
      score_heading_angles.push_back(angle_score);
      score_heading_velos.push_back(velo_score);
      score_obstacles.push_back(obs_score);
      valid_paths.push_back(path);
    }
    else
    {
      // 無効なパスはスキップ
      continue;
    }
  }

  if (valid_paths.empty())
  {
    std::cerr << "No valid paths found. All paths are either out of bounds or colliding with obstacles." << std::endl;
    throw std::runtime_error("有効なPathが存在しません。全てのPathがコース外か障害物と衝突しています。");
  }

  // スコアの正規化
  std::vector<double> score_heading_angles_norm = minMaxNormalize(score_heading_angles);
  std::vector<double> score_heading_velos_norm = minMaxNormalize(score_heading_velos);
  std::vector<double> score_obstacles_norm = minMaxNormalize(score_obstacles);

  // ゴールへの距離を計算
  std::vector<double> distances_to_goal;
  for (const auto &path : valid_paths)
  {
    double last_x = path.getX().back();
    double last_y = path.getY().back();
    double distance = std::hypot(g_x - last_x, g_y - last_y);
    distances_to_goal.push_back(distance);
  }

  // 距離を反転（近いほど良い）して正規化
  std::vector<double> inverted_distances;
  for (const auto &d : distances_to_goal)
  {
    inverted_distances.push_back(-d);
  }
  std::vector<double> distances_normalized = minMaxNormalize(inverted_distances);

  // 最適なpathを探索
  double max_score = -std::numeric_limits<double>::infinity();
  int opt_index = -1;
  std::vector<double> opt_cost_components(4, 0.0); // 角度, 速度, 障害物, 距離

  for (size_t k = 0; k < valid_paths.size(); ++k)
  {
    double temp_score = (params_.WEIGHT_ANGLE * score_heading_angles_norm[k] +
                         params_.WEIGHT_VELOCITY * score_heading_velos_norm[k] +
                         params_.WEIGHT_OBSTACLE * score_obstacles_norm[k] +
                         params_.WEIGHT_DISTANCE * distances_normalized[k]);
    if (temp_score > max_score)
    {
      max_score = temp_score;
      opt_index = static_cast<int>(k);
      opt_cost_components[0] = score_heading_angles_norm[k];
      opt_cost_components[1] = score_heading_velos_norm[k];
      opt_cost_components[2] = score_obstacles_norm[k];
      opt_cost_components[3] = distances_normalized[k];
    }
  }

  Path opt_path = valid_paths[opt_index];

  // 最適パスのコスト情報を出力
  std::cout << "選択されたパスのコスト:" << std::endl;
  std::cout << "  角度スコア: " << opt_cost_components[0] << std::endl;
  std::cout << "  速度スコア: " << opt_cost_components[1] << std::endl;
  std::cout << "  障害物スコア: " << opt_cost_components[2] << std::endl;
  std::cout << "  ゴール距離スコア: " << opt_cost_components[3] << std::endl;

  return {opt_path, valid_paths};
}

double DWA::headingAngle(const Path &path, double g_x, double g_y) const
{
  double last_x = path.getX().back();
  double last_y = path.getY().back();
  double last_th = path.getTh().back();

  // double angle_to_goal = std::atan2(g_y - last_y, g_x - last_x);
  // double score_angle = angle_to_goal - last_th;
  // score_angle = std::abs(angleRangeCorrector(score_angle));
  // score_angle = M_PI - score_angle;

  // return score_angle;

  // 目標への角度を計算
  double angle_to_goal = std::atan2(g_y - last_y, g_x - last_x);
  
  // 現在の向きとの角度差を計算
  double angle_diff = angle_to_goal - last_th;

  // 角度を正規化
  angle_diff = std::atan2(std::sin(angle_diff), std::cos(angle_diff)); // -πからπの範囲に正規化

  // スコアを計算（角度差が小さいほど良い）
  double score_angle = 1.0 - std::abs(angle_diff) / M_PI; // 0から1の範囲にスケーリング

  return score_angle;
}

double DWA::headingVelo(const Path &path) const
{
  // return path.getUV();

  double velocity = path.getUV();
  double max_velocity = params_.MAX_SPEED; // 最大速度を取得

  // スコアを計算（速度が高いほど良い）
  double score_velocity = velocity / max_velocity; // 0から1の範囲にスケーリング

  // スコアが1を超えないように制限
  return std::min(score_velocity, 1.0);
}

std::vector<Obstacle> DWA::calcNearestObs(const Robot &state, const std::vector<Obstacle> &obstacles) const
{
  double area_dis_to_obs = 30.0; // パラメーター
  std::vector<Obstacle> nearest_obs;
  for (const auto &obs : obstacles)
  {
    double temp_dis = std::hypot(state.getX() - obs.getX(), state.getY() - obs.getY());
    if (temp_dis < area_dis_to_obs)
    {
      nearest_obs.push_back(obs);
    }
  }
  return nearest_obs;
}

double DWA::obstacleScore(const Path &path, const std::vector<Obstacle> &nearest_obs) const
{
  double score_obstacle = 2.1;
  for (size_t i = 0; i < path.getX().size(); ++i)
  {
    for (const auto &obs : nearest_obs)
    {
      double temp_dis = std::hypot(path.getX()[i] - obs.getX(), path.getY()[i] - obs.getY());
      if (temp_dis < obs.getSize())
      {
        return -std::numeric_limits<double>::infinity(); // 無効なパス
      }
      if (temp_dis < score_obstacle)
      {
        score_obstacle = temp_dis;
      }
    }
  }
  return score_obstacle;
}

const std::vector<std::vector<Path>> &DWA::getTrajPaths() const
{
  return traj_paths_;
}

const std::vector<Path> &DWA::getTrajOpt() const
{
  return traj_opt_;
}

const std::vector<double> &DWA::getTrajGX() const
{
  return traj_g_x_;
}

const std::vector<double> &DWA::getTrajGY() const
{
  return traj_g_y_;
}

const Course &DWA::getCourse() const
{
  return course_;
}

// サイズ管理関数の作成
void DWA::addTrajectory(Path opt_path, double g_x, double g_y, const std::vector<Path> &valid_paths)
{
  const size_t MAX_TRAJ_SIZE = 10;
  if (traj_opt_.size() >= MAX_TRAJ_SIZE)
  {
    traj_opt_.erase(traj_opt_.begin());
  }
  traj_opt_.push_back(opt_path);

  if (traj_g_x_.size() >= MAX_TRAJ_SIZE)
  {
    traj_g_x_.erase(traj_g_x_.begin());
  }
  traj_g_x_.push_back(g_x);

  if (traj_g_y_.size() >= MAX_TRAJ_SIZE)
  {
    traj_g_y_.erase(traj_g_y_.begin());
  }
  traj_g_y_.push_back(g_y);

  if (traj_paths_.size() >= MAX_TRAJ_SIZE)
  {
    traj_paths_.erase(traj_paths_.begin());
  }
  traj_paths_.push_back(valid_paths);
}

File: aichallenge/workspace/src/aichallenge_submit/dwa/dwa/dwa/dwa.py
import math
import numpy as np
from utils import min_max_normalize, angle_range_corrector
from obstacle import Obstacle
from config import *
from course import Course
import sys
from config import LOOKAHEAD_DISTANCE

class Simulator_DWA_robot:
    def __init__(self):
        self.max_accelation = MAX_ACCEL
        self.max_ang_accelation = MAX_DYAWRATE
        self.lim_max_velo = MAX_SPEED
        self.lim_min_velo = MIN_SPEED
        self.lim_max_ang_velo = MAX_YAWRATE
        self.lim_min_ang_velo = MIN_YAWRATE

    def predict_state(self, ang_velo, velo, x, y, th, dt, pre_step):
        next_xs = []
        next_ys = []
        next_ths = []

        for _ in range(pre_step):
            temp_x = velo * math.cos(th) * dt + x
            temp_y = velo * math.sin(th) * dt + y
            temp_th = ang_velo * dt + th

            next_xs.append(temp_x)
            next_ys.append(temp_y)
            next_ths.append(temp_th)

            x = temp_x
            y = temp_y
            th = temp_th

        return next_xs, next_ys, next_ths

class Path:
    def __init__(self, u_th, u_v):
        self.x = None
        self.y = None
        self.th = None
        self.u_v = u_v
        self.u_th = u_th


class DWA():
    def __init__(self):
        self.simu_robot = Simulator_DWA_robot()
        self.pre_time = PREDICT_TIME
        self.pre_step = int(self.pre_time / DT)
        self.delta_velo = V_RESOLUTION
        self.delta_ang_velo = YAWRATE_RESOLUTION
        self.samplingtime = DT
        self.weight_angle = WEIGHT_ANGLE
        self.weight_velo = WEIGHT_VELOCITY
        self.weight_obs = WEIGHT_OBSTACLE
        self.weight_distance = WEIGHT_DISTANCE
        try:
            self.course = Course(LEFT_LANE_BOUND_FILE, RIGHT_LANE_BOUND_FILE, CENTER_LANE_LINE_FILE)
        except FileNotFoundError as e:
            print("CSVファイルのパスが正しいか確認してください。")
            sys.exit(1)

        # すべてのPathを保存
        self.traj_paths = []
        self.traj_opt = []

    def calc_input(self, robot, obstacles):
        # Path作成
        paths = self._make_path(robot)
        print(f"Number of paths generated: {len(paths)}")

        # Path評価
        g_x, g_y = self.course.get_next_target_point(robot.x, robot.y, robot.th, LOOKAHEAD_DISTANCE)

        opt_path, valid_paths = self._eval_path(paths, g_x, g_y, robot, obstacles)

        self.traj_opt.append(opt_path)
        if not hasattr(self, 'traj_g_x'):
            self.traj_g_x = []
            self.traj_g_y = []
        self.traj_g_x.append(g_x)
        self.traj_g_y.append(g_y)

        self.traj_paths.append(valid_paths)  # 有効なパスのみを保存

        return valid_paths, opt_path

    def _make_path(self, state):
        # 角度と速度の範囲算出
        min_ang_velo, max_ang_velo, min_velo, max_velo = self._calc_range_velos(state)
        max_velo = MAX_SPEED
        # 全てのpathのリスト
        paths = []

        # 角速度と速度の組み合わせを全探索
        for ang_velo in np.arange(min_ang_velo, max_ang_velo, self.delta_ang_velo):
            for velo in np.arange(min_velo, max_velo, self.delta_velo):

                path = Path(ang_velo, velo)

                next_x, next_y, next_th \
                    = self.simu_robot.predict_state(ang_velo, velo, state.x, state.y, state.th, self.samplingtime, self.pre_step)

                path.x = next_x
                path.y = next_y
                path.th = next_th

                # 作ったpathを追加
                paths.append(path)

        # 時刻歴Pathを保存
        self.traj_paths.append(paths)

        return paths

    def _calc_range_velos(self, state): # 角速度と角度の範囲決定①
        # 角速度
        range_ang_velo = self.samplingtime * self.simu_robot.max_ang_accelation
        min_ang_velo = state.u_th - range_ang_velo
        max_ang_velo = state.u_th + range_ang_velo
        # 最小値
        if min_ang_velo < self.simu_robot.lim_min_ang_velo:
            min_ang_velo = self.simu_robot.lim_min_ang_velo
        # 最大値
        if max_ang_velo > self.simu_robot.lim_max_ang_velo:
            max_ang_velo = self.simu_robot.lim_max_ang_velo

        # 速度
        range_velo = self.samplingtime * self.simu_robot.max_accelation
        min_velo = state.u_v - range_velo
        max_velo = state.u_v + range_velo
        # 最小値
        if min_velo < self.simu_robot.lim_min_velo:
            min_velo = self.simu_robot.lim_min_velo
        # 最大値
        if max_velo > self.simu_robot.lim_max_velo:
            max_velo = self.simu_robot.lim_max_velo

        return min_ang_velo, max_ang_velo, min_velo, max_velo

    def _eval_path(self, paths, g_x, g_y, state, obstacles):
        # 一番近い障害物判定
        nearest_obs = self._calc_nearest_obs(state, obstacles)
        valid_paths = []
        score_heading_angles = []
        score_heading_velos = []
        score_obstacles = []

        # 全てのpathで評価を検索
        for path in paths:
            # (1) heading_angle
            angle_score = self._heading_angle(path, g_x, g_y)
            # (2) heading_velo
            velo_score = self._heading_velo(path)
            # (3) obstacle
            obs_score = self._obstacle(path, nearest_obs)
            # パスが有効かチェック（スコアにinfやnanが含まれていないか）
            if np.isfinite(angle_score) and np.isfinite(velo_score) and np.isfinite(obs_score):
                score_heading_angles.append(angle_score)
                score_heading_velos.append(velo_score)
                score_obstacles.append(obs_score)
                valid_paths.append(path)
            else:
                # 無効なパスはスキップ
                continue

        if not valid_paths:
            print("No valid paths found. All paths are either out of bounds or colliding with obstacles.")
            raise ValueError("有効なPathが存在しません。全てのPathがコース外か障害物と衝突しています。")

        # スコアの正規化
        score_heading_angles = min_max_normalize(score_heading_angles)
        score_heading_velos = min_max_normalize(score_heading_velos)
        score_obstacles = min_max_normalize(score_obstacles)

        # ゴールへの距離を計算
        distances_to_goal = []
        for path in valid_paths:
            last_x = path.x[-1]
            last_y = path.y[-1]
            distance = math.hypot(g_x - last_x, g_y - last_y)
            distances_to_goal.append(distance)

        # 距離を反転（近いほど良い）して正規化
        inverted_distances = [-d for d in distances_to_goal]
        distances_normalized = min_max_normalize(inverted_distances)

        # 最適なpathを探索
        score = -float('inf')  # スコアの初期値を負の無限大に設定
        opt_path = None        # opt_path を初期化
        opt_cost_components = {}

        for k in range(len(valid_paths)):
            temp_score = (self.weight_angle * score_heading_angles[k] +
                          self.weight_velo * score_heading_velos[k] +
                          self.weight_obs * score_obstacles[k] +
                          self.weight_distance * distances_normalized[k])  # 距離スコアを加算

            if temp_score > score:
                opt_path = valid_paths[k]
                score = temp_score
                opt_cost_components = {
                    'angle_score': score_heading_angles[k],
                    'velo_score': score_heading_velos[k],
                    'obs_score': score_obstacles[k],
                    'distance_score': distances_normalized[k]
                }

        if opt_path is None:
            # フォールバック: 有効なパスがあれば最初を選択
            if valid_paths:
                opt_path = valid_paths[0]
            else:
                raise ValueError("Pathが生成されていません。")

        # 最適パスのコスト情報を出力
        print("選択されたパスのコスト:")
        print(f"  角度スコア: {opt_cost_components.get('angle_score', 'N/A')}")
        print(f"  速度スコア: {opt_cost_components.get('velo_score', 'N/A')}")
        print(f"  障害物スコア: {opt_cost_components.get('obs_score', 'N/A')}")
        print(f"  ゴール距離スコア: {opt_cost_components.get('distance_score', 'N/A')}")

        return opt_path, valid_paths  # 有効なパスも返すように変更

    def _heading_angle(self, path, g_x, g_y): # ゴールに向いているか
        # 終端の向き
        last_x = path.x[-1]
        last_y = path.y[-1]
        last_th = path.th[-1]

        # 角度計算
        angle_to_goal = math.atan2(g_y-last_y, g_x-last_x)

        # score計算
        score_angle = angle_to_goal - last_th

        # ぐるぐる防止
        score_angle = abs(angle_range_corrector(score_angle))

        # 最大と最小をひっくり返す
        score_angle = math.pi - score_angle

        return score_angle

    def _heading_velo(self, path): # 速くんでいるか（直進）

        score_heading_velo = path.u_v

        return score_heading_velo

    def _calc_nearest_obs(self, state, obstacles):
        area_dis_to_obs = 10 # パラメー（何メートル考慮するか，本当は制動距離）
        nearest_obs = [] # あるエリアに入ってる障害物
        for obs in obstacles:
            temp_dis_to_obs = math.sqrt((state.x - obs.x) ** 2 + (state.y - obs.y) ** 2)

            if temp_dis_to_obs < area_dis_to_obs :
                nearest_obs.append(obs)

        return nearest_obs

    def _obstacle(self, path, nearest_obs):
        # 障害物回避（エリアに入ったらその線は使わない）/ (障害物ともっとも近い距離)))
        score_obstacle = 2.1
        temp_dis_to_obs = 0.0

        for i in range(len(path.x)):
            for obs in nearest_obs:
                temp_dis_to_obs = math.sqrt((path.x[i] - obs.x) ** 2 + (path.y[i] - obs.y) ** 2)

                if temp_dis_to_obs < obs.size + 1.0:
                    return -float('inf')  # 無効なパスを示すスコアを返す

                if temp_dis_to_obs < score_obstacle:
                    score_obstacle = temp_dis_to_obs  # 一番近いところ

        return score_obstacle









